Source/WebCore/xml/XSLTUnicodeSort.cpp

```diff
@@ -31,117 +31,96 @@
 
 #if ENABLE(XSLT)
 
+#include <algorithm>
 #include <libxslt/templates.h>
 #include <libxslt/xsltutils.h>
 #include <wtf/Vector.h>
 #include <wtf/unicode/Collator.h>
 
-WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
-
 namespace WebCore {
 
 // Based on default implementation from libxslt 1.1.22 and xsltICUSort.c example.
-void xsltUnicodeSortFunction(xsltTransformContextPtr ctxt, xmlNodePtr *sorts, int nbsorts)
+void xsltUnicodeSortFunction(xsltTransformContextPtr ctxt, xmlNodePtr* rawSorts, int nbsorts)
 {
-#ifdef XSLT_REFACTORED
-    xsltStyleItemSortPtr comp;
-#else
-    const xsltStylePreComp* comp;
-#endif
-    xmlXPathObjectPtr *resultsTab[XSLT_MAX_SORT];
-    xmlXPathObjectPtr *results = NULL, *res;
-    xmlNodeSetPtr list = NULL;
-    int len = 0;
-    int i, j, incr;
-    int tst;
-    int depth;
-    xmlNodePtr node;
-    xmlXPathObjectPtr tmp;    
-    int number[XSLT_MAX_SORT], desc[XSLT_MAX_SORT];
-
-    if ((ctxt == NULL) || (sorts == NULL) || (nbsorts <= 0) ||
-        (nbsorts >= XSLT_MAX_SORT))
+    if (!ctxt || !rawSorts || nbsorts <= 0 || nbsorts >= XSLT_MAX_SORT)
         return;
-    if (sorts[0] == NULL)
+
+    auto sorts = unsafeMakeSpan(rawSorts, nbsorts);
+    if (!sorts[0])
         return;
-    comp = static_cast<xsltStylePreComp*>(sorts[0]->psvi);
-    if (comp == NULL)
+
+    auto comp = static_cast<xsltStylePreComp*>(sorts[0]->psvi);
+    if (!comp)
         return;
 
-    list = ctxt->nodeList;
-    if ((list == NULL) || (list->nodeNr <= 1))
+    auto list = ctxt->nodeList;
+    if (!list || list->nodeNr <= 1)
         return; /* nothing to do */
 
-    for (j = 0; j < nbsorts; j++) {
+    std::array<int, XSLT_MAX_SORT> desc;
+    std::array<int, XSLT_MAX_SORT> number;
+    for (size_t j = 0; j < sorts.size(); ++j) {
         comp = static_cast<xsltStylePreComp*>(sorts[j]->psvi);
-        if ((comp->stype == NULL) && (comp->has_stype != 0)) {
-            xmlChar* stype =
-                xsltEvalAttrValueTemplate(ctxt, sorts[j], (const xmlChar *) "data-type", XSLT_NAMESPACE);
+        if (!comp->stype && comp->has_stype) {
+            auto* stype = xsltEvalAttrValueTemplate(ctxt, sorts[j], (const xmlChar *) "data-type", XSLT_NAMESPACE);
```

Comment:
Why not use reinterpret_cast for xmlChar*?

Reply:
Actually, Darin introduced `byteCast<>()` for such proposes. I'll make the change.

Reply:
Actually `byteCast<>()` doesn't seem to work here, I'll use `reinterpret_cast`.

Reply:
I seem to recall I had trouble making `byteCast` work on string literals, which is a shame because the whole point is that it’s safer than `reinterpret_cast`, which can do almost anything!